src/modules/aws_opensearch/mappers/domain.ts

```diff
@@ -0,0 +1,398 @@
+    Object.is(a.coldStorage, b.coldStorage) &&
+    Object.is(a.dedicatedMasterType, b.dedicatedMasterType) &&
+    Object.is(a.dedicatedMasterCount, b.dedicatedMasterCount) &&
+    Object.is(a.autoTune, b.autoTune) &&
+    Object.is(a.subnets?.length, b.subnets?.length) &&
+    !!a.subnets?.every(sa => !!b.subnets?.find(sb => Object.is(sa.id, sb.id))) &&
+    Object.is(a.securityGroups?.length, b.securityGroups?.length) &&
+    !!a.securityGroups?.every(sa => !!b.securityGroups?.find(sb => Object.is(sa.id, sb.id))) &&
+    Object.is(a.enableFineGrainedAccessControl, b.enableFineGrainedAccessControl) &&
+    Object.is(a.fineGrainedAccessControlUserArn, b.fineGrainedAccessControlUserArn) &&
+    Object.is(a.fineGrainedAccessControlMasterUsername, b.fineGrainedAccessControlMasterUsername) &&
+    Object.is(a.fineGrainedAccessControlMasterPassword, b.fineGrainedAccessControlMasterPassword) &&
+    policiesAreSame(a.accessPolicy, b.accessPolicy) &&
+    Object.is(a.endpoint, b.endpoint);
+
+  private async waitForUpdate(client: OpenSearch, domainName: string) {
+    await createWaiter<OpenSearch, DescribeDomainChangeProgressCommandInput>(
+      {
+        client,
+        // all in seconds
+        maxWaitTime: 1800,
+        minDelay: 1,
+        maxDelay: 4,
+      },
+      { DomainName: domainName },
+      async (cl, cmd) => {
+        const response = await cl.describeDomainChangeProgress(cmd);
+        if (!response.ChangeProgressStatus) return { state: WaiterState.RETRY };
+        if (response.ChangeProgressStatus.Status === 'COMPLETED') return { state: WaiterState.SUCCESS };
+        else if (['PENDING', 'PROCESSING'].includes(response.ChangeProgressStatus.Status ?? ''))
+          return { state: WaiterState.RETRY };
+        else if (response.ChangeProgressStatus.Status === 'FAILED') return { state: WaiterState.FAILURE };
+
+        return { state: WaiterState.RETRY };
+      },
+    );
+
+    await createWaiter<OpenSearch, DescribeDomainsCommandInput>(
+      {
+        client,
+        // all in seconds
+        maxWaitTime: 300,
+        minDelay: 1,
+        maxDelay: 4,
+      },
+      { DomainNames: [domainName] },
+      async (cl, cmd) => {
+        const data = await cl.describeDomains(cmd);
+        if (!!data.DomainStatusList?.[0].Endpoint && data.DomainStatusList?.[0].Processing === false)
+          return { state: WaiterState.SUCCESS };
+        else return { state: WaiterState.RETRY };
+      },
+    );
+  }
+
+  private async domainMapper(domain: AwsDomain, region: string, ctx: Context) {
+    const out: Domain = new Domain();
+    if (domain.Deleted) return undefined;
+
+    out.domainName = domain.DomainName!;
+    if (domain.DomainEndpointOptions?.CustomEndpoint) {
+      const certificateId = awsAcmModule.certificate.generateId({
+        arn: domain.DomainEndpointOptions?.CustomEndpointCertificateArn ?? '',
+      });
+      out.customEndpoint = domain.DomainEndpointOptions.CustomEndpoint;
```

Comment:
do custom endpoints only need to have a custom certificate?

Reply:
Yes. Only custom endpoints use it.

Reply:
and is it a must to have a certificate associated?

Reply:
Yes. 
<img width="789" alt="image" src="https://user-images.githubusercontent.com/5357435/228872504-c540b253-4b94-4e92-a33d-5242904bbeca.png">